Make switching to coupled defaults easier

WalterKolczynski-NOAA opened the pull request:
In order to facilitate running coupled experiments without having to
modify several config files, some updates are made to the config files.

A new file, config.defaults.s2sw is created to contain the alternate
defaults for the coupled model, based on prototype 7. This file is
sourced whenever the app is S2SW.

The UFS app is made into a variable that is now set by the setup scripts
via the --app switch. Allowed values are ATM, ATMW, S2S, S2SW. If no
argument is provided, the default is ATM (so it doesn't change existing
behavior).

The section in config.base that sets all the components based on the app
is moved earlier in order to allow the new defaults to be set.

Settings in some config files have been modifyied to now accept values
that have been set previously to allow the alternate settings for the
coupled model. In the future, this could be extended to additional config
files to allow more settings to be overridden.

waveGRDN and waveGRDG are updated to be based on waveGRD.

Updated instructions to setup prototype analog:

1. Clone and run checkout, build_all, and link as normal for coupled:
```
./checkout.sh
./build_all.sh -c
./link_workflow emc orion coupled
```

2. In ush/rocoto, run
```
./setup_expt_fcstonly.py --app S2SW --res 384 --pslot $PSLOT --comrot $COMROT --expdir $EXPDIR --idate 2013040100 --edate 2013040100 --configdir ../../parm/config
./setup_workflow_fcstonly.py --expdir $EXPDIR/$PSLOT
```

3. In $EXPDIR/$PSLOT, modify $PSLOT.xml as follows:
- Remove or comment out the entire gfsinit task
- Remove or comment out gfsinit as a dependency for gfswaveprep
- Optionally reduce the walltime for the forecast job to 1 hour to move through the queue faster
-- The forecast job uses too many nodes for the debug queue, though you can use the debug queue for the wave prep jobs

4. In $ROTDIR, do the following:
```
cd $ROTDIR
ln -s /work/noaa/global/wkolczyn/noscrub/global-workflow/FV3ICS FV3ICS

mkdir -p $ROTDIR/$PSLOT/gfs.20130401/00/atmos
cd $ROTDIR/$PSLOT/gfs.20130401/00/atmos
ln -s $ROTDIR/FV3ICS/2013040100/atmos/C384/INPUT INPUT

mkdir -p $ROTDIR/$PSLOT/gdas.20130331/18/wave/restart
cd $ROTDIR/$PSLOT/gdas.20130331/18/wave/restart
ln -s $ROTDIR/FV3ICS/2013040100/wav/20130401.000000.restart.gwes_30m .
```

Refs: #368

WalterKolczynski-NOAA committed:
Make switching to coupled defaults easier

In order to facilitate running coupled experiments without having to
modify several config files, some updates are made to the config files.

A new file, config.defaults.s2sw is created to contain the alternate
defaults for the coupled model, based on prototype 7. This file is
sourced whenever the app is S2SW.

The UFS app is made into a variable that is now set by the setup scripts
via the --app switch. Allowed values are ATM, ATMW, S2S, S2SW. If no
argument is provided, the default is ATM (so it doesn't change existing
behavior).

The section in config.base that sets all the components based on the app
is moved earlier in order to allow the new defaults to be set.

Settings in some config files have been modifyied to now accept values
that have been set previously to allow the alternate settings for the
coupled model. In the future, this could be extended to additional config
files to allow more settings to be overridden.

waveGRDN and waveGRDG are updated to be based on waveGRD.

Updated instructions to setup prototype analog:

1. Clone and run checkout, build_all, and link as normal for coupled:
```
./checkout.sh
./build_all.sh -c
./link_workflow emc orion coupled
```

2. In ush/rocoto, run
```
./setup_expt_fcstonly.py --app S2SW --res 384 --pslot $PSLOT --comrot $COMROT --expdir $EXPDIR --idate 2013040100 --edate 2013040100 --configdir ../../parm/config
./setup_workflow_fcstonly.py --expdir $EXPDIR/$PSLOT
```

3. In $EXPDIR/$PSLOT, modify $PSLOT.xml as follows:
- Remove or comment out the entire gfsinit task
- Remove or comment out gfsinit as a dependency for gfswaveprep
- Optionally reduce the walltime for the forecast job to 1 hour to move through the queue faster
-- The forecast job uses too many nodes for the debug queue, though you can use the debug queue for the wave prep jobs

4. In $ROTDIR, do the following:
```
cd $ROTDIR
ln -s /work/noaa/global/wkolczyn/noscrub/global-workflow/FV3ICS FV3ICS

mkdir -p $ROTDIR/$PSLOT/gfs.20130401/00/atmos
cd $ROTDIR/$PSLOT/gfs.20130401/00/atmos
ln -s $ROTDIR/FV3ICS/2013040100/atmos/C384/INPUT INPUT

mkdir -p $ROTDIR/$PSLOT/gdas.20130331/18/wave/restart
cd $ROTDIR/$PSLOT/gdas.20130331/18/wave/restart
ln -s $ROTDIR/FV3ICS/2013040100/wav/20130401.000000.restart.gwes_30m .
```

Refs: #368

WalterKolczynski-NOAA commented:
Produces bitwise-identical results for the coupled model. Going to test atm-only now.

aerorahul commented on parm/config/config.fcst:
```diff
@@ -93,45 +93,45 @@ export new_o3forc="YES"
-export satmedmf=".true."
-export isatmedmf=1
-tbf=""
+export hybedmf=${hybedmf:-".false."}
```
`variable=${variable:-default_value}`
implies one can define this variable before this script.
This being a configuration script then looses the point of being a configuration script, rather become a configurable script.

The syntax at this level should remain at key-value pairs.

WalterKolczynski-NOAA replied:
This was (to me) the least objectionable method to handle multiple classes of default settings:
- If we source the coupled defaults afterwards, you have to worry about also redoing any calculations that were made based on the original defaults, plus you run into that problem where a user will change a setting in one place, only to not have it do anything because it is being overwritten elsewhere.
- If we insert different defaults where they appear, you wind up with a bunch of if/switch blocks based on the app scattered all over the config files.
- We already agreed keeping an additional set of config files was a bad idea, because of the bloat and duplication

This method also has the advantage of being able to be easily extended more broadly to something sorta like the case files from the CROW workflow, where the config files _would_ basically become a configurable script where you just have one file where you make any overrides. Even if we don't do that (since we're building a new system), this way is the cleanest option in the interim, IMHO. Unless there is an option I'm missing?

We can discuss it in the tag-up tomorrow if need be.

WalterKolczynski-NOAA replied:
Per our conversation, all of the overrides except those necessary for the alternate settings were removed.

aerorahul reviewed (changes requested):
The config scripts should remain `key=value`.

KateFriedman-NOAA commented on parm/config/config.defaults.s2sw:
```diff
@@ -0,0 +1,37 @@
+# config.fcst
+IALB=2
+IEMS=2
+DO_CA="YES"
```
What does the "CA" in `DO_CA` stand for?

WalterKolczynski-NOAA replied:
Cellular Automata

KateFriedman-NOAA commented on parm/config/config.fv3:
```diff
@@ -42,110 +42,110 @@ fi
-        export WRITE_GROUP_GFS=1
-        export WRTTASK_PER_GROUP_GFS=$npe_node_max
-        export WRTIOBUF="1M"
+        export DELTIM=${DELTIM:-1200}
```
Why have overrides in this config? This is where these values get set...so what would potentially override these values?

WalterKolczynski-NOAA replied:
Prototype 7 is using a different timestep and layout than listed here for C384

KateFriedman-NOAA replied:
Gotcha. Should we just pull these values right from wherever those values are set then? Skip the middle man. Or is that the next step in moving towards the model template files?

WalterKolczynski-NOAA replied:
See the above response to Rahul. Once you start chopping it up, you're going to have if or switch blocks all over the place where these settings pop up to pick the wrong ones.

KateFriedman-NOAA commented on ush/rocoto/setup_expt.py:
```diff
@@ -135,6 +136,7 @@ def edit_baseconfig():
 link initial condition files from $ICSDIR to $COMROT'''
 
     parser = ArgumentParser(description=description, formatter_class=ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--app', help='UFS application', type=str, choices=['ATM', 'ATMW', 'S2S', 'S2SW'], required=False, default='ATM')
```
Not request for a change, just a thought...I imagine this list of possible APPs will show itself in multiple places, is there a way to consolidate all of that? For example, the value of `choices` here could be sourced from some sort of common file/place earlier in the script. Have been contemplating something similar for a while to handle the list of supported machines/target that's listed/used through the system and other similar lists that will grow/shrink over time as we support more/less. Food for thought, thanks!

WalterKolczynski-NOAA replied:
Right now it is just three places (the two py scripts and config.base). And in config.base, it takes the form of case statements that might not be wise to abstract. The two py scripts have a LOT of duplicated code (it probably should be just a single script with fcst-only and DA options); app choices aren't unique in that respect.

KateFriedman-NOAA reviewed (commented):
Changes look good from a technical standpoint...but want to discuss the addition of overrides in the configs (see my comment in the review). Love the addition of the `--app` argument, thanks for adding that!

WalterKolczynski-NOAA commented:
ATM-only fcst-only completed successfully.

JessicaMeixner-NOAA commented on parm/config/config.wave:
```diff
@@ -26,28 +26,29 @@ export CDUMPRSTwave="gdas"
+export waveGRD=${waveGRD:-'gnh_10m aoc_9km gsh_15m'}
+nGRD=$(echo $waveGRD | wc -w)
+export waveGRDN=$(seq $nGRD)               # gridnumber for ww3_multi
+export waveGRDG=$(seq -f "%g0" $nGRD)      # gridgroup for ww3_multi
```
While this works for this setting, this theoretically is not the formula for all sets of grids, so I actually don't think we should put this formula in.

JessicaMeixner-NOAA replied:
Here's a way I started to make this easier for configuring waves but never finished testing: 
https://github.com/JessicaMeixner-NOAA/global-workflow/blob/coupled-sprint/runwithwaves/parm/config/config.wave#L97-L171   I don't know if this or what you have is preferable.  They both work, I think we just need to set waveGRDN, waveGRDG in your parm/config/config.defaults.s2sw file

WalterKolczynski-NOAA replied:
Backed this change out. I'd still love to have a way to auto-generate these if possible in the future.

JessicaMeixner-NOAA replied:
You cannot autogenerate them because they are specific to the combination of grids you choose and how you are running them.

WalterKolczynski-NOAA committed:
Undo addition of all but necessary overrides

Removes most of the setting overrides from config files except
those currently necessary to run prototype from defaults file.

Also backed out the change that automatically calcluated the wave
grid and group numbers.

Refs: #368

JessicaMeixner-NOAA reviewed (approved):
Thanks for updating the wave variables.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -180,15 +233,15 @@ export gfs_cyc=@gfs_cyc@ # 0: no GFS cycle, 1: 00Z only, 2: 00Z and 12Z only, 4:
-export FHMAX_GFS_06=384
-export FHMAX_GFS_12=384
-export FHMAX_GFS_18=384
+export FHMAX_GFS_00=${FHMAX_GFS_00:-384}
+export FHMAX_GFS_06=${FHMAX_GFS_06:-384}
+export FHMAX_GFS_12=${FHMAX_GFS_12:-384}
+export FHMAX_GFS_18=${FHMAX_GFS_18:-384}
```
I have always wondered, is there ever a practical reason why a user would set different forecast lengths for different cycles?

WalterKolczynski-NOAA replied:
I believe there have been times in the past when 00z or 00z & 12z have been run longer than other cycles.

KateFriedman-NOAA replied:
Yeah it's a legacy-type thing to let users run different lengths for different cycles. From a technical standpoint I dislike this and would prefer to merge this back into a single forecast length variable and just let users define which cycles to run gfs instead of having different forecast lengths per cycle which messes with the verification.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -180,15 +233,15 @@ export gfs_cyc=@gfs_cyc@ # 0: no GFS cycle, 1: 00Z only, 2: 00Z and 12Z only, 4:
-export FHMAX_HF_GFS=0
-export FHOUT_HF_GFS=1
+export FHOUT_GFS=${FHOUT_GFS:-3}
+export FHMAX_HF_GFS=${FHMAX_HF_GFS:-0}
+export FHOUT_HF_GFS=${FHOUT_HF_GFS:-1}
```
I think the current develop of the UFS now no longer has `HF` as an input.
I think they want all forecast hours explicitly defined as a list.
Please check though.

WalterKolczynski-NOAA replied:
This is true, but we're currently still using these settings to build the output list later in the workflow: https://github.com/NOAA-EMC/global-workflow/blob/feature/coupled_sprint/ush/forecast_predet.sh#L57-L68 (this was in the PR I made last week to update the UFS version).

aerorahul reviewed (approved):
Looks good.
Thanks for taking in the suggestions.
Also, see a couple of comments.
